components/server/src/workspace/workspace-deletion-service.ts

```diff
@@ -94,8 +99,79 @@ export class WorkspaceDeletionService {
      * @param ws
      * @param includeSnapshots
      */
-    protected async deleteWorkspaceStorage(ws: WorkspaceAndOwner, includeSnapshots: boolean): Promise<boolean> {
-        await this.storageClient.deleteWorkspaceBackups(ws.ownerId, ws.id, includeSnapshots);
+    protected async deleteWorkspaceStorage(
+        ctx: TraceContext,
+        ws: WorkspaceAndOwner,
+        includeSnapshots: boolean,
+    ): Promise<boolean> {
+        const span = TraceContext.startSpan("deleteWorkspaceStorage", ctx);
+        try {
+            await this.storageClient.deleteWorkspaceBackups(ws.ownerId, ws.id, includeSnapshots);
+            let vss = await this.db.trace({ span }).findVolumeSnapshotForGCByWorkspaceId(ws.id);
+            await Promise.all(vss.map((vs) => this.garbageCollectVolumeSnapshot({ span }, vs)));
+        } catch (err) {
+            TraceContext.setError({ span }, err);
+            throw err;
+        } finally {
+            span.finish();
+        }
         return true;
     }
+
+    /**
+     * Perform deletion of volume snapshot from all clusters and from gloud provider:
+     *  - throws an error if something went wrong during deletion
+     *  - returns true in case of successful deletion
+     * @param ctx
+     * @param vs
+     */
+    public async garbageCollectVolumeSnapshot(ctx: TraceContext, vs: VolumeSnapshot): Promise<boolean> {
+        const span = TraceContext.startSpan("garbageCollectVolumeSnapshot", ctx);
+
+        try {
+            const allClusters = await this.workspaceManagerClientProvider.getAllWorkspaceClusters();
+            // we need to do two things here:
+            // 1. we want to delete volume snapshot object from all workspace clusters
+            // 2. we want to delete cloud provider source snapshot
+            let wasDeleted = false;
+            let index = 0;
+
+            let availableClusters = allClusters.filter((c) => c.state === "available");
+            for (let cluster of availableClusters) {
```

Comment:
I expected this method to connect to a specific ws-manager based on a "region" field, either retrieved from `VolumeSnapshot`, `Workspace` or `WorkspaceInstance`. Iterating over all currently connected clients makes it impossible to detect cases where we are no longer connected to a certain ws-manager.

Or is this due to the fact that we might have multiple ws-manager (with different names) in the same GCloud region, and we do not know how to map those? :thinking:

Reply:
Not quite. Imagine following scenario:
EU cluster where WS was started from Volume Snapshot X
US cluster where WS was also started from same Volume Snapshot X (from prebuild for example)

Then we want to GC that volume snapshot.
We need to do two things now:
1. Ensure that we delete k8s object representing this snapshot from ALL clusters (to make sure we do not accumulate those over time, as it will affect k8s api server)
2. Ensure we delete volume snapshot object from cloud provider

So this code ensures that. That is why it attempts to talk to all existing clusters and clean it up from all of them. 
If we are no longer connected to some cluster, I think that is fine, as I would assume that cluster is getting decommissioned.

Reply:
> We need to do two things now:

Ah, that explains it. Just to ensure we're on the same page: `ws-manager` does _not_ remove the PVC object from the control plane when the corresponding workspace is stopped?

> Ensure we delete volume snapshot object from cloud provider
I thought we just need to do this, and also wondered if this was something we put into content-service...? :thinking:

Reply:
Sorry to "block" the merging of the PR here, but I feel I'm still missing some context to understand why the PVC lifecycle looks the way it does.

Reply:
> Just to ensure we're on the same page: `ws-manager` does _not_ remove the PVC object from the control plane when the corresponding workspace is stopped?

No, when the workspace stopped, the PVC object would be removed after the VolumeSnapshot object is ready to use. (Which means the PVC snapshot is done).

> > Ensure we delete volume snapshot object from cloud provider
> 
> I thought we just need to do this, and also wondered if this was something we put into content-service...? 🤔

The lifecycle of PersistentVolumeClaim/VolumeSnapshot is managed by ws-manager.

Reply:
@geropl 
>Just to ensure we're on the same page: ws-manager does not remove the PVC object from the control plane when the corresponding workspace is stopped?

it does. There are two objects in play here. PVC is an actual storage object, it is only alive while workspace is running. When workspace is stopped, wsman will "convert" PVC into VolumeSnapshot object that is stored in cloud provider (outside of k8s control).

When we open workspace, PVC is created and restored from VolumeSnapshot object.

When it is time to delete\GC volume snapshot, we might have same VolumeSnapshot across multiple clusters (prebuild that was opened in different clusters for example). So we want to clean up that volume snapshot object from all clusters (as otherwise we will be accumulating a lot of them over the lifecycle of the cluster), and then also remove actual snapshot from cloud provider as well.

All this work is done by wsman currently.

It will be better to move that into content service indeed, but currently it is not possible, because content service still lives in webapp cluster, and should be migrated into workspace clusters first.

Reply:
Thx @sagor999 for syncing! :+1: 
I understood that:
 - workspace takes care of deleting the actual [disk snapshots](https://cloud.google.com/compute/docs/disks/snapshots?hl=en)
 - we keep the cluster-local VolumeSnapshots and VolumeSnapshotContent in the cluster to optimize for fast workspace starts
 - but to avoid cluttering the clusters too much, we aim to remove those here whenever their disk snapshot get deleted

It would be nice if we added some comments along those lines to the GC. :pray: